controllers/mirrorpeer_controller.go

```diff
@@ -118,54 +123,59 @@ func (r *MirrorPeerReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		mirrorPeerCopy.Labels[hubRecoveryLabel] = "resource"
 		err = r.Client.Update(ctx, mirrorPeerCopy)
 
-		if k8serrors.IsConflict(err) {
-			logger.Info("MirrorPeer is being updated by another process. Retrying", "MirrorPeer", mirrorPeerCopy.Name)
-			return ctrl.Result{Requeue: true}, nil
-		} else if k8serrors.IsNotFound(err) {
-			logger.Info("MirrorPeer no longer exists. Ignoring since object must have been deleted", "MirrorPeer", mirrorPeerCopy.Name)
-			return ctrl.Result{}, nil
-		} else if err != nil {
-			logger.Info("Warning: Failed to update mirrorpeer", "MirrorPeer", mirrorPeerCopy.Name, "Error", err)
-		} else {
-			logger.Info("Successfully added label to mirrorpeer for disaster recovery", "MirrorPeer", mirrorPeerCopy.Name)
-			return ctrl.Result{Requeue: true}, nil
+		if err != nil {
+			return utils.CheckK8sUpdateErrors(err, mirrorPeerCopy)
 		}
+		logger.Info("Successfully added label to mirrorpeer for disaster recovery", "Mirrorpeer", mirrorPeerCopy.Name)
+		return ctrl.Result{Requeue: true}, nil
 	}
 
 	if mirrorPeer.Status.Phase == "" {
 		mirrorPeer.Status.Phase = multiclusterv1alpha1.ExchangingSecret
 		statusErr := r.Client.Status().Update(ctx, &mirrorPeer)
 		if statusErr != nil {
 			logger.Error(statusErr, "Error occurred while updating the status of mirrorpeer", "MirrorPeer", mirrorPeer)
-			// Requeue, but don't throw
 			return ctrl.Result{Requeue: true}, nil
 		}
 	}
-	// Create or Update ManagedClusterAddon
-	for i := range mirrorPeer.Spec.Items {
-		annotations := make(map[string]string)
-		annotations[utils.DRModeAnnotationKey] = string(mirrorPeer.Spec.Type)
-		managedClusterAddOn := addonapiv1alpha1.ManagedClusterAddOn{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:        tokenexchange.TokenExchangeName,
-				Namespace:   mirrorPeer.Spec.Items[i].ClusterName,
-				Annotations: annotations,
-			},
+	if mirrorPeer.GetDeletionTimestamp().IsZero() {
+		if !utils.ContainsString(mirrorPeer.GetFinalizers(), mirrorPeerFinalizer) {
+			logger.Info("Finalizer not found on MirrorPeer. Adding Finalizer")
+			mirrorPeer.Finalizers = append(mirrorPeer.Finalizers, mirrorPeerFinalizer)
+			if err := r.Client.Update(ctx, &mirrorPeer); err != nil {
+				logger.Info("Failed to add finalizer to MirrorPeer")
+				return utils.CheckK8sUpdateErrors(err, &mirrorPeer)
+			}
```

Comment:
This should be checked before setting labels for recovery.
Then, recovery label and finalizer canbe added via same call.